src/scheduler/node_info.c

```diff
@@ -240,52 +303,104 @@ query_nodes(int pbs_sd, server_info *sinfo)
 	}
 
 	if ((ninfo_arr = (node_info **) malloc((num_nodes + 1) * sizeof(node_info *))) == NULL) {
-		log_err(errno, "query_nodes", "Error allocating memory");
+		log_err(errno, __func__, MEM_ERR_MSG);
 		pbs_statfree(nodes);
 		return NULL;
 	}
 	ninfo_arr[0] = NULL;
 
 #ifdef NAS /* localmod 049 */
 	if ((sinfo->nodes_by_NASrank = (node_info **) malloc(num_nodes * sizeof(node_info *))) == NULL) {
-		log_err(errno, "query_nodes", "Error allocating nodes_by_NASrank memory");
+		log_err(errno, __func__, MEM_ERR_MSG);
 		pbs_statfree(nodes);
 		free_nodes(ninfo_arr);
 		return NULL;
 	}
 #endif /* localmod 049 */
 
-	cur_node = nodes;
-	for (i = 0, nidx = 0; cur_node != NULL; i++) {
-		/* get node info from server */
-		if ((ninfo = query_node_info(cur_node, sinfo)) == NULL) {
-			pbs_statfree(nodes);
-			free_nodes(ninfo_arr);
-			return NULL;
+	j = 0;
+	chunk_size = num_nodes / num_threads;
+	chunk_size = (chunk_size > 1024) ? chunk_size : 1024;
+	while (num_nodes > 0) {
```

Comment:
I really wanted to do that as well, but it won't really reduce much code, and it will introduce many if else conditions as each task's data structure is different, has different members etc., so I think that one function will end up being a mess.

Reply:
Right, how about we have one common function to place work and notify threads and then one function to dequeue results back. Currently, there is a lot of repetition in many places.

Reply:
I'm talking about that function only Arun. Just to give you an idea, that one function's argument list will look something like this:

void send_work(int ct_left, enum thread_task_type task_type, node_info **onodes, node_info **nnodes, server_info *sinfo, unsigned int flags, struct batch_status *nodes, resource_resv **oresresv_arr, resource_resv **nresresv_arr, queue_info *nqinfo ..... (so on)

and this will need to be expanded as we add more tasks. The function body will similarly be quite a mess.

I can put the following 4 lines in a function, so I'll do that, please let me know if you can think of a clean way to generalize anything else.
pthread_mutex_lock(&work_lock);
ds_enqueue(work_queue, (void *) task);
pthread_cond_signal(&work_cond);
pthread_mutex_unlock(&work_lock);